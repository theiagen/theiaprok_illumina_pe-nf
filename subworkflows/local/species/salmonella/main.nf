include { SISTR     } from '../../../../modules/local/species_typing/salmonella/sistr/main'
include { SEQSERO2  } from '../../../../modules/local/species_typing/salmonella/seqsero2/main'
include { GENOTYPHI } from '../../../../modules/local/species_typing/salmonella/genotyphi/main'

workflow SALMONELLA_SPECIES_TYPING {

    take:
    ch_salmonella // Channel of tuples [meta, assembly, reads, species]

    main:

    ch_versions = Channel.empty()
    ch_sistr_results = Channel.empty()
    ch_seqsero2_results = Channel.empty()
    ch_genotyphi_results = Channel.empty()

    // Extract assembly for SISTR
    ch_assembly = ch_salmonella.map { meta, assembly, reads, species -> [meta, assembly] }
    
    // Extract reads for SeqSero2 (when needed)
    ch_reads = ch_salmonella.map { meta, assembly, reads, species -> [meta, reads] }

    // Run SISTR on assembly
    SISTR (
        ch_assembly
    )
    ch_sistr_results = SISTR.out.sistr_result
    ch_versions = ch_versions.mix(SISTR.out.versions)
    
    // SeqSero2 - different input based on data type
    if (params.ont_data || params.assembly_only) {
        SEQSERO2 (
            ch_assembly
        )
    } else {
        SEQSERO2 (
            ch_reads
        )
    }
    ch_seqsero2_results = SEQSERO2.out.seqsero2_report
    ch_versions = ch_versions.mix(SEQSERO2.out.versions)
    
    // GenotypHi for Typhi - only if not assembly_only
    if (!params.assembly_only) {
        // Get serotype predictions to check for Typhi
        ch_typhi_check = SEQSERO2.out.seqsero_serotype
            .join(SISTR.out.sistr_predicted_serotype)
            .filter { meta, seqsero_file, sistr_file ->
                def seqsero_content = seqsero_file.text.trim()
                def sistr_content = sistr_file.text.trim()
                seqsero_content == "Typhi" || sistr_content == "Typhi"
            }
            .map { meta, seqsero_file, sistr_file -> meta }
        
        ch_typhi_reads = ch_reads
            .join(ch_typhi_check)
            .map { meta, reads, check -> [meta, reads] }
        
        GENOTYPHI (
            ch_typhi_reads
        )
        ch_genotyphi_results = GENOTYPHI.out.genotyphi_report
        ch_versions = ch_versions.mix(GENOTYPHI.out.versions)
    }

    emit:
    sistr_results = ch_sistr_results
    seqsero2_results = ch_seqsero2_results
    genotyphi_results = ch_genotyphi_results
    
    versions = ch_versions
}