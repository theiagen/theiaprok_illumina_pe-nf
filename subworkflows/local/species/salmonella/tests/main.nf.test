nextflow_workflow {
    name "Test Subworkflow SALMONELLA_SPECIES"
    script "../main.nf"
    workflow "SALMONELLA_SPECIES_TYPING"

    test("salmonella_species - salmonella characterization") {
        setup {
            run("GUNZIP") {
                script "../../../../../modules/nf-core/gunzip/main.nf"
                process {
                    """
                    input[0] = Channel.of([
                        [id:'test_salmonella', single_end:false],
                        file('https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/006/945/GCF_000006945.2_ASM694v2/GCF_000006945.2_ASM694v2_genomic.fna.gz', checkIfExists: true)
                    ])
                    """
                }
            }
        }
        
        when {
            workflow {
                """
                input[0] = GUNZIP.out.gunzip
                    .map { meta, assembly -> 
                        [
                            meta, 
                            assembly, 
                            [
                                file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                                file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                            ],
                            "Salmonella"
                        ]
                    }
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.sistr_results,
                    workflow.out.seqsero2_results,
                    workflow.out.genotyphi_results,
                    workflow.out.versions
                ).match() }
            )
        }
    }
}