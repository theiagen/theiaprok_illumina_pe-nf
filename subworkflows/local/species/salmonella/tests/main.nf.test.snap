{
    "salmonella_species - salmonella characterization": {
        "content": [
            [
                [
                    {
                        "id": "test_salmonella",
                        "single_end": false
                    },
                    "test_salmonella.tsv:md5,987a5643f44a7b3c27f4959db8193155"
                ]
            ],
            [
                [
                    {
                        "id": "test_salmonella",
                        "single_end": false
                    },
                    "test_salmonella_SeqSero_result.tsv:md5,b2a31c7c25608c1d34ca46a695e68d47"
                ]
            ],
            [
                
            ],
            [
                "versions.yml:md5,19f2910e118e21ac364fa6f3376d9797",
                "versions.yml:md5,c181548419c7b6ef88b0f52f75cf9ac6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-07-02T20:21:40.669418902"
    }
}